## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [200]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [201]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [202]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [203]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [204]:
# What three item categories had the most entries?
df.value_counts('category', ascending=False)

#DFQ - consumables, furniture, software


category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [205]:
# For the category with the most entries, which subcategory had the most entries? BATHROOM SUPPLIES!
#So, I need to isolate the category that occurs most often (consumables), then do a value count of the number of unique occurrences in that category

cat_subcat_count = df[['category', 'subcategory']]
subcat_count = cat_subcat_count.value_counts('subcategory')
subcat_count


subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
seating               2664
pens                  2603
logistics             1122
telecom                978
audio visual           732
cabling                655
desks                  510
computers              467
Name: count, dtype: int64

In [206]:
# Which five clients had the most entries in the data?REYES, EVERETT, MYERS, YOUNG, GARRETT
top_clients = df['client_id'].value_counts()
top_five_clients = top_clients.nlargest(5)
top_five_clients




client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [207]:
# Store the client ids of those top 5 clients in a list.
top_five_clients.to_csv('top.five.clients.csv', index=True, header=True)


In [208]:
# How many total units (the qty column) did the client with the most entries order order?

top_client_id = df['client_id'].value_counts().idxmax()
top_client_filtered = df[df['client_id'] == top_client_id]
top_client_qty = top_client_filtered['qty'].sum()
top_client_qty


64313

In [209]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [210]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df_part_two = df
df_part_two.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [211]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
#if unit_weight > 50, then charge $7; if unit_weight < 50, then charge $10

# Calculate total weight for each row
total_weight = df_part_two['unit_weight'] * df_part_two['qty']

# Apply the shipping price based on the condition
df_part_two['shipping_price'] = total_weight.apply(lambda x: 7 if x > 50 else 10)
df_part_two['shipping_price'] = (df_part_two['shipping_price'] * df_part_two['unit_weight'])
new_df = ['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'shipping_price', 'subtotal']
df_part_two.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0           52.50  
1           14.90  
2           11.76  
3           12.30  
4          325.01

In [212]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

df_part_two['shipping_price'] = df_part_two['shipping_price'].astype(float)

df_part_two['total_price'] = (df_part_two['subtotal'] + df_part_two['shipping_price']) * 1.0925
df_part_two['total_price'] = df_part_two['total_price'].round(2)

df_part_two.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00           52.50   
1      15.09         1.49   21            0     523.95           14.90   
2       7.86         1.68   39            6     527.28           11.76   
3      24.85         1.23   29            3    1056.18           12.30   
4     108.17        46.43   20            1    3902.00          325.01   

   total_price  
0    125874.03  
1       588.69  
2       588.90  
3      1167.31  
4      4618.01  

[5 rows x 21 columns]

In [213]:
# Create a column for the cost of each line using unit cost, qty, and shipping price (assume the shipping cost is exactly what is charged to the client).

# Create a column for the cost of each line. Use unit cost, qty, and ...
# In this case "line" is synonymous with "record" or "row". Better wording altogether:
# Create a column that calculates total cost (including shipping) for each row.

df_part_two['line_cost'] = df_part_two['unit_cost'] * df_part_two['qty'] + df_part_two['shipping_price']
df_part_two.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price  total_price  \
0         7.50  105            1  115164.00           52.50    125874.03   
1         1.49   21            0     523.95           14.90       588.69   
2         1.68   39            6     527.28           11.76       588.90   
3         1.23   29            3    1056.18           12.30      1167.31   
4        46.43   20            1    3902.00          325.01      4618.01   

   line_cost  
0   80137.05  
1     331.79  
2     318.30  
3     732.95  
4    2488.41  

[5 rows x 22 columns]

In [214]:
# Create a column for the profit of each line using line cost and line price

df_part_two['line_profit'] = df_part_two['total_price'] - df_part_two['line_cost']
df_part_two.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price  total_price  line_cost  line_profit  
0            1  115164.00           52.50    125874.03   80137.05     45736.98  
1            0     523.95           14.90       588.69     331.79       256.90  
2            6     527.28           11.76       588.90     318.30       270.60  
3            3    1056.18           12.30      1167.31     732.95       434.36  
4            1    3902.00          325.01      4618.01    2488.41      2129.60  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [215]:
# Check your work using the totals above
check_work = df_part_two.loc[
                            (df_part_two['order_id'] == 2742071) |
                            (df_part_two['order_id'] == 2173913) |
                            (df_part_two['order_id'] == 6128929)]
check_work_df = check_work
check_work_df.groupby('order_id')['total_price'].agg(sum)

#groupby easch order number, then add the total price values



C:\Users\Daniel Quinn\AppData\Local\Temp\ipykernel_21084\1228972172.py:7: FutureWarning: The provided callable <built-in function sum> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  check_work_df.groupby('order_id')['total_price'].agg(sum)


order_id
2173913    119760.03
2742071    139791.74
6128929    752608.13
Name: total_price, dtype: float64

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [216]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# need to group by client_id, qty
# then multiply qty*unit_price
#sort descending
top_clients = df['client_id'].value_counts()
top_five_clients = top_clients.nlargest(5)
top_five_clients

top_five_spend = df.groupby('client_id')['qty']

df['qty_spend'] = df['qty']*df['unit_price']
df_filtered = df[df['client_id'].isin(top_five_clients.index)]
total_spend = df_filtered.groupby('client_id')['qty_spend'].sum()
total_spend.head()

# for client in top_five_clients.index:
#     #print(client)

#     client_spend = df.loc[df['client_id'] == client]
#     client_spend['qty_spend'] = client_spend['qty']*client_spend['unit_price']
#     #print(client_spend['qty_spend'])

# client_spend = client_spend.groupby('client_id')['qty_spend'].sum()
# print(client_spend)



client_id
24741    70176885.25
33615     5839032.11
38378     8384321.15
46820     7317356.03
66037     7995708.38
Name: qty_spend, dtype: float64

In [217]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

df_part_two['Total Profit (millions)'] = df_part_two['line_profit']
df_part_two.sort_values(by='Total Profit (millions)', ascending=False)
df_part_two.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   shipping_price  total_price  line_cost  line_profit  qty_spend  \
0           52.50    125874.03   80137.05     45736.98  115164.00   
1           14.90       588.69     331.79       256.90     523.95   
2           11.76       588.90     318.30       270.60     527.28   
3           12.30      1167.31     732.95       434.36    1056.18   
4          325.01      4618.01    2488.41      2129.60    3902.00   

   Total Profit (millions)  
0                 45736.98  
1                   256.90  
2                   270.60  
3                   434.36  
4                  2129.60  

[5 rows x 25 columns]

In [218]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.

n_top_five_clients = df_part_two['client_id'].value_counts().nlargest(5).index
filtered_df = df_part_two[df_part_two['client_id'].isin(n_top_five_clients)]
filtered_df

almost_final_df = filtered_df[['client_id', 'qty', 'shipping_price', 'total_price', 'line_profit']]
almost_final_df

grouped_df = almost_final_df.groupby('client_id').sum().reset_index()
grouped_df

#filtered_df = df_part_two[df_part_two['client_id'].isin(top_five_clients)]
#filtered_df

#final_data_top_five = final_data.sort_values(by = 'qty', ascending = False)
#final_data_top_five


#df_part_two_summary = pd.DataFrame()

#top_five_clients[columns = Total Units Purchased, Total Shipping Price, Total Revenue, and Total Profit]


client_id     qty  shipping_price  total_price  line_profit
0      24741  239862         6741.02  76675611.68  36106419.32
1      33615   64313         8958.67   6388929.94   2033642.25
2      38378   73667         8433.12   9169084.10   2955385.40
3      46820   75768         7895.03   8002836.77   2589198.94
4      66037   43018         7159.32   8743132.96   3126642.51

In [219]:
# Format the data and rename the columns to names suitable for presentation.
grouped_df = grouped_df.rename(columns=
                               {'client_id': 'Client ID',
                                'qty': 'Total Units Purchased',
                                'shipping_price': 'Total Shipping Price',
                                'total_price': 'Total Revenue',
                                'line_profit': 'Total Profit'})

# Define the money columns. 


# Define a function that converts a dollar amount to millions.
   

# Apply the currency_format_millions function to only the money columns. 

# Rename the columns to reflect the change in the money format. 
grouped_df['Total Shipping Price'] = grouped_df['Total Shipping Price'].apply(lambda x: f'${x:,.2f}')
grouped_df['Total Revenue'] = grouped_df['Total Revenue'].apply(lambda x: f'${x:,.2f}')
grouped_df['Total Profit'] = grouped_df['Total Profit'].apply(lambda x: f'${x:,.2f}')

grouped_df

Client ID  Total Units Purchased Total Shipping Price   Total Revenue  \
0      24741                 239862            $6,741.02  $76,675,611.68   
1      33615                  64313            $8,958.67   $6,388,929.94   
2      38378                  73667            $8,433.12   $9,169,084.10   
3      46820                  75768            $7,895.03   $8,002,836.77   
4      66037                  43018            $7,159.32   $8,743,132.96   

     Total Profit  
0  $36,106,419.32  
1   $2,033,642.25  
2   $2,955,385.40  
3   $2,589,198.94  
4   $3,126,642.51